src/common/snippets/src/pass/propagate_precision.cpp

```diff
@@ -0,0 +1,277 @@
+
+    std::unordered_map<std::shared_ptr<ngraph::opset1::Result>, element::Type> result_types;
+    auto results = f->get_results();
+    for (auto& result : results) {
+        result_types.emplace(result, result->get_input_source_output(0).get_element_type());
+    }
+
+    for (const auto& op : f->get_ordered_ops()) {
+        auto type_info = op->get_type_info();
+        if (!target_machine->has(type_info)) {
+            throw ov::Exception(
+                "operation '" + std::string(type_info.version_id) + "::" + std::string(type_info.name) +
+                "' was not found in target machine");
+        }
+
+        auto exec = target_machine->get_supported_precisions(type_info);
+        const auto supported_precisions = exec(op);
+        if (supported_precisions.empty()) {
+            continue;
+        }
+
+        // There are two operation types which break precision propagation:
+        //   1) Existing convertion operations. Solution: remove convertion
+        //      operation before general algo
+        //   2) Type relaxed based operations. Will be resolved by snippet opset.
+
+        auto input_precisions_were_changed = false;
+
+        for (const auto& input : op->inputs()) {
+            const auto convert = ngraph::as_type<snippets::op::ConvertSaturation>(input.get_source_output().get_node());
+            if (convert == nullptr) {
+                continue;
+            }
+
+            const auto precision_before = convert->input(0).get_source_output().get_element_type();
+            const auto precision_after = convert->output(0).get_element_type();
+            if (can_be_removed(precision_before, precision_after, precision_before)) {
+                op->set_argument(input.get_index(), convert->input(0).get_source_output());
+                input_precisions_were_changed = true;
+            }
+        }
+
+        std::vector<element::Type> input_precisions;
+        for (const auto& input : op->inputs()) {
+            const auto input_precision = input.get_source_output().get_element_type();
+            input_precisions.push_back(input_precision);
+        }
+
+        assert(std::all_of(
+            supported_precisions.begin(),
+            supported_precisions.end(),
+            [&input_precisions](const std::vector<element::Type>& precisions) {
+                return precisions.size() == input_precisions.size();
+            }) && "input precisions count is not equal for supported precisions");
+
+        // update input precisions
+        // if possible then convert precisions to supported
+        if (!supported_precisions.empty() &&
+            !std::any_of(
+                supported_precisions.begin(),
+                supported_precisions.end(),
+                [&input_precisions](const std::vector<element::Type>& precisions) {
+                    return precisions == input_precisions;
+                })) {
```

Comment:
Optional: we could use `std::all_of` here instead of `!std::any_of` (and change functor). As for me, it's more readable

Reply:
done